### PR TITLE
Refactor indent/crystal.vim

### PR DESCRIPTION
**Pull request by jlcrochet**
Recently, I've been running into a lot of bugs related to indentation that have become increasingly difficult to fix. I took a look at vim-ruby's `indent` and liked what they've done with it recently, so I decided that the easiest way to continue maintaining vim-crystal's `indent` is to refactor it to more closely match the current state of vim-ruby.

In addition to making the `indent` code more organized, this refactor has allowed me to fix some regressions and edge cases that were not being handled properly before. Unfortunately, it also means that this is going to be another large PR; I apologize for that. Additionally, there are some edge cases that I was unable to come up with an efficient solution for that will have to be addressed at some other time.

I made a few other minor changes as well, so I'll summarize everything I did below:

* `indent`
  * Refactored to match the layout of vim-ruby's `indent`. A lot of code has been moved to `autoload/crystal/indent.vim`.
  * Two new configuration variables have been inherited from vim-ruby: `g:crystal_indent_assignment_style` and `g:crystal_indent_block_style`.
    ```crystal
    # g:crystal_indent_assignment_style = "hanging" (default)
    foo = if true
            5
          else
            10
          end

    # g:crystal_indent_assignment_style = "variable"
    foo = if true
      5
    else
      10
    end
    ```

    ```crystal
    # g:crystal_indent_block_style = "expression" (default)
    xs.each do |x|
      puts x
    end

    # g:crystal_indent_block_style = "do"
    xs.each do |x|
              puts x
            end
    ```
* `ftplugin`
  * Removed some group names from `b:match_skip` that no longer exist.
* `syntax`
  * Removed the `defined?` method, since it doesn't exist in Crystal.
  * Fixed highlighting for multiline `{{ ... }}` macro tags.

### CODE REVIEW

**Comment by jlcrochet** [comment]
Currently working on some more bugfixes and optimizations; will likely mark as ready for review sometime tomorrow.

File: ftplugin/crystal.vim [hunk 1]
```diff
@@ -56,12 +31,12 @@ command! -buffer -nargs=? CrystalSpecRunCurrent call crystal_lang#run_current_sp
...
-nnoremap <buffer><Plug>(crystal-spec-switch)        :<C-u>CrystalSpecSwitch<CR>
-nnoremap <buffer><Plug>(crystal-spec-run-all)       :<C-u>CrystalSpecRunAll<CR>
-nnoremap <buffer><Plug>(crystal-spec-run-current)   :<C-u>CrystalSpecRunCurrent<CR>
-nnoremap <buffer><Plug>(crystal-format)             :<C-u>CrystalFormat<CR>
+nnoremap <buffer><Plug>(crystal-jump-to-definition) <Cmd><C-u>CrystalDef<CR>
+nnoremap <buffer><Plug>(crystal-show-context)       <Cmd><C-u>CrystalContext<CR>
+nnoremap <buffer><Plug>(crystal-spec-switch)        <Cmd><C-u>CrystalSpecSwitch<CR>
+nnoremap <buffer><Plug>(crystal-spec-run-all)       <Cmd><C-u>CrystalSpecRunAll<CR>
+nnoremap <buffer><Plug>(crystal-spec-run-current)   <Cmd><C-u>CrystalSpecRunCurrent<CR>
+nnoremap <buffer><Plug>(crystal-format)             <Cmd><C-u>CrystalFormat<CR>
```
Comment (rhysd) [inline ftplugin/crystal.vim]: `<Cmd>` only works on Neovim. Please keep `:` here.

**Comment by jlcrochet** [comment]
Since this PR adds `g:crystal_indent_assignment_style`, it should fix #42.

Additionally, db697ba adds support for the `annotation` keyword, which fixes #57.

**Comment by rhysd** [comment]
I'm sorry for the delay. I'll take look at this tomorrow.

**Comment by jlcrochet** [comment]
Actually, I'm currently doing some additional work on another branch at the moment. I'll mark this as ready for review again once it's done.

File: autoload/crystal/indent.vim [hunk 2]
```diff
@@ -1,162 +1,885 @@
+let s:cpo_save = &cpo
+set cpo&vim
```
Comment (rhysd) [inline autoload/crystal/indent.vim]: I think cpo-save is no longer necessary. It was necessary for ancient Vim or setting `compatible` option (please see `:help use-cpo-save`). But with such an ancient Vim, this plugin would not work (since other features are also not supported). `compatible` option is as well.

File: indent/crystal.vim [hunk 3]
```diff
@@ -1,9 +1,26 @@
...
 " Only load this indent file when no other was loaded.
 if exists('b:did_indent')
   finish
 endif
+
 let b:did_indent = 1
 
+if !exists('g:crystal_indent_assignment_style')
+  " Possible values: 'variable', 'hanging'
+  let g:crystal_indent_assignment_style = 'hanging'
```
Comment (rhysd) [inline indent/crystal.vim]: Would you check behavior of `crystal tool format` for default value of this variable? I guess behavior of `crystal tool format` is "variable".

File: indent/crystal.vim [hunk 4]
```diff
@@ -1,9 +1,26 @@
...
+" ==============
+
 " Only load this indent file when no other was loaded.
 if exists('b:did_indent')
   finish
 endif
+
 let b:did_indent = 1
 
+if !exists('g:crystal_indent_assignment_style')
```
Comment (rhysd) [inline indent/crystal.vim]: Please move initialization of `g:crystal_*` variables to `plugin/crystal.vim`.

File: syntax/crystal.vim [hunk 5]
```diff
@@ -243,53 +244,56 @@ SynFold '<<' syn region crystalString start=+\%(\%(class\|::\)\_s*\|\%([]}).]\)\
...
 " Keywords
 " Note: the following keywords have already been defined:
 " begin case class def do end for if module unless until while
 syn match crystalControl        "\<\%(break\|next\|rescue\|return\)\>[?!]\@!"
-syn match crystalOperator       "\<defined?" display
 syn match crystalKeyword        "\<\%(super\|previous_def\|yield\|of\|with\|uninitialized\|union\)\>[?!]\@!"
 syn match crystalBoolean        "\<\%(true\|false\)\>[?!]\@!"
-syn match crystalPseudoVariable "\<\%(nil\|self\|__DIR__\|__FILE__\|__LINE__\|__END_LINE__\)\>[?!]\@!" " TODO: reorganise
+syn match crystalPseudoVariable "\<\%(nil\|__DIR__\|__FILE__\|__LINE__\|__END_LINE__\)\>[?!]\@!" " TODO: reorganise
+syn match crystalPseudoVariable "\<self\>"
```
Comment (rhysd) [inline syntax/crystal.vim:274]: Would you tell me why this change is necessary? Are `self!` and `self?` permitted in Crystal?

**Review by rhysd — CHANGES_REQUESTED** [review]

**Comment by rhysd** [comment]
> so I decided that the easiest way to continue maintaining vim-crystal's indent is to refactor it to more closely match the current state of vim-ruby.

I agree with this direction.

File: syntax/crystal.vim (hunk 5 above)
Comment (jlcrochet) [reply syntax/crystal.vim:274]: This was to allow for `self` to be properly higlighted as a nilable type in type declarations. For example:

```crystal
struct Enum
  def self.parse?(string : String) : self?
    ...
  end
end
```

File: autoload/crystal/indent.vim (hunk 2 above)
Comment (jlcrochet) [reply autoload/crystal/indent.vim]: For now, I have removed all references to cpo_save.

File: ftplugin/crystal.vim (hunk 1 above)
Comment (jlcrochet) [reply ftplugin/crystal.vim]: Sorry, I wasn't aware of that.

File: indent/crystal.vim (hunk 3 above)
Comment (jlcrochet) [reply indent/crystal.vim]: According to my tests, it appears that the default values inherited from vim-ruby are the same as `crystal tool format`, so I don't think any change is necessary here.

Comment (rhysd) [reply indent/crystal.vim]: I see. Let's keep 'hanging' as default value.

File: syntax/crystal.vim (hunk 5 above)
Comment (rhysd) [reply syntax/crystal.vim:274]: Thanks for the explanation. I didn't know that becuase I haven't used `self` as type.

**Review by rhysd — APPROVED** [review]
At this point it looks good for me.

**Comment by jlcrochet** [comment]
I'm currently working on another commit that will add a few performance optimizations, if you don't mind reviewing it. I should be able to finish it up tomorrow.

File: syntax/crystal.vim (hunk 5 above)
Comment (jlcrochet) [reply syntax/crystal.vim:274]: Reference for how to use `self` as a type, just in case you didn't know: https://crystal-lang.org/reference/syntax_and_semantics/type_restrictions.html#self-restriction.

**Comment by rhysd** [comment]
Sure. Please remove WIP when this PR is ready. I'll review it again.

**Comment by jlcrochet** [comment]
Sorry for the delay on this: it's taking a lot longer than I thought, and I'm running out of time to work on it for now. It may be another week or two before I can get it finished.

**Comment by rhysd** [comment]
> Sorry for the delay on this: it's taking a lot longer than I thought, and I'm running out of time to work on it for now. It may be another week or two before I can get it finished.

Please take your pace. Since this is refactoring, we don't need to rush.

**Comment by jlcrochet** [comment]
After extensive experimentation, I've ended up rewriting large parts of both the indent and syntax files; this was mostly to correct numerous performance issues that I found in the code inherited from vim-ruby. However, as I continue to make such changes, the code is going to become increasingly difficult for you to review and maintain and for others to understand; additionally, I will probably be making changes to things like syntax highlighting that make sense for me, but not for other people. Thus, I've decided to relegate future changes to my own fork for personal use. I will continue to submit bugfixes for other users to this repo as needed.

Since I won't be making any further changes to this PR, it can be merged now if it looks good to you.

**Comment by rhysd** [comment]
@jlcrochet 

OK, let me check this (since changes are a bit big, it may take time). Please resolve the current conflicts.

**Comment by jlcrochet** [comment]
By the way, I haven't committed any additional changes since 99fefb8, which I believe has already been reviewed.

**Comment by rhysd** [comment]
Yeah, let me check again for sure.

**Review by rhysd — APPROVED** [review]
